File: two_pointers/minimum_consecutive_cards_to_pick_up.rb
# @param {Integer[]} cards
# @return {Integer}
def minimum_card_pickup(cards)
    window = Hash.new(0)
    shortest = cards.size + 1
    left = 0
    
    cards.each_with_index do |card, right|
        window[card] += 1
        while window[card] == 2
            shortest = [shortest, right - left + 1].min
            window[cards[left]] -= 1
            left += 1
        end
    end
    return shortest == cards.size + 1 ? -1 : shortest
end

File: two_pointers/minimum_consecutive_cards_to_pick_up.py
from collections import defaultdict

class Solution:
    def minimumCardPickup(self, cards: List[int]) -> int:
        window = defaultdict(int)
        shortest = len(cards) + 1
        left = 0

        for right, card in enumerate(cards):
            window[card] += 1
            while window[card] == 2:
                shortest = min(shortest, right - left + 1)
                window[cards[left]] -= 1
                left += 1

        return -1 if shortest == len(cards) + 1 else shortest
